advtraj/integrate/forward.py

```diff
@@ -1,25 +1,575 @@
+    the origin at x(n).
+
+    Parameters
+    ----------
+    ds_position_scalars : Xarray DataSet
+        3D fields of trajectory tracer data.
+    ds_traj_posn_org : Xarray DataSet
+        Current trajectory positions.
+    ds_traj_posn_first_guess : Xarray DataSet
+        First guess of trajectory endpoints.
+    interpolator : fast interp interpolator.
+        Pre-calculated interpolator for fields in ds_position_scalars.
+        If None, interpolator is calculated on each call to the interpolation.
+    solver : str
+        'PI' or 'PI_hybrid'.
+    interp_order : int, optional
+        Interpolation order. The default is 5.
+    pioptions : dict, optional
+        Options for solution and minimization. The default is None.
+        pioptions available::
+
+            maxiter (int)        : Maximum number of iterations. Delault=500.
+            miniter (int)        : Number of iterations to use before adjusting
+                                   relax. Default=10.
+            relax (float)        : Initial relaxation factor. Default=0.8.
+            relax_reduce (float) : Factor to mutiply relax after every miniter
+                                   iterations. Default=0.9.
+            tol (float)          : Distance in terms of grid lengths defining
+                                   convergence. Default=0.01.
+            opt_one_step (bool)  : If True, use single call to optimizer for
+                                   all trajectories. Default=False.
+            disp (bool)          : Print information on convergence.
+                                   Default=False.
+            norm (str)           : Norm to use in erro. Options are
+                                   'mean_abs_error', 'max_abs_error'
+                                   or RMS distance. Default='max_abs_error'
+            minoptions (dict)    : Options for minimiser. Default:
+                minoptions = {'max_outer_loops': 1, 'solver':'BFGS',
+                            'minimize_options':{'maxiter': 50,'disp': False,}}
+
+    Returns
+    -------
+    Xarray DataSet, numpy array [3, n]
+        New trajectory position and residual error vector.
+
+    """
+    options = {}
+    if pioptions is not None:
+        options.update(pioptions)
+
+    # defaults
+    maxiter = options.get("maxiter", 500)
+    miniter = options.get("miniter", 10)
+    relax = options.get("relax", 0.8)
+    relax_reduce = options.get("relax_reduce", 0.9)
+    tol = options.get("tol", 0.01)
+    opt_one_step = options.get("opt_one_step", False)
+    disp = options.get("disp", False)
+    norm = options.get("norm", "max_abs_error")
+    # = options.get('', )
+
+    grid_spacing = np.array(
+        [
+            ds_position_scalars["x"].attrs["dx"],
+            ds_position_scalars["y"].attrs["dy"],
+            ds_position_scalars["z"].attrs["dz"],
+        ]
+    )
+
+    grid_spacing = grid_spacing[:, np.newaxis]
+
+    def get_error_norm(
```

Comment:
I would split `get_error_norm` out of this function. You could make `norm` an argument to that function

Reply:
Done.